### ISSUE TITLE
release wheels for python 3.9

### ISSUE THREAD

**Issue by xmatthias** [PyTables/PyTables#823]
Currently, [pypi](https://pypi.org/project/tables/#files) does not contain wheels for python 3.9.

I know python 3.9 just came out - but it would be great if the last release (or a new release - considering the last release has been over a year ago) could be updated to include Wheels for python 3.9 to simplify installation.

**Comment by amotl** [comment]
Dear @avalentino, @tomkooij, @PyTables and @numfocus,

first things first: Thank you so much for conceiving and maintaining PyTables.

Binary wheels for Python 3.9 for all of lxml, h5py, Pandas and Numpy are already available on PyPI these days. At @earthobservations, we just have been able to upgrade to the most recent releases on [Wetterdienst](https://github.com/earthobservations/wetterdienst) and it works like a charm, see https://github.com/earthobservations/wetterdienst/pull/202#issuecomment-727190718.

It would be awesome if you could also build and upload `cp39` artefacts to PyPI to make it easier for downstream packages and their users to upgrade to Python 3.9.

With kind regards,
Andreas.

**Comment by avalentino** [comment]
OK i managed to build wheels for linux locally.
@tomkooij  if you do not have plans to build wheels via [pytables-wheels](https://github.com/MacPython/pytables-wheels) I could try to upload mines.

**Comment by amotl** [comment]
Dear Antonio,

thanks a bunch already. While I am just getting started to learn about the package publishing infrastructure behind the different Python packages for scientific computing, there is nothing that I could really contribute here other than pointing out what others are doing, especially in the light of switching away from Rackspace and staging packages through either AppVeyor/Travis/Anaconda or GitHub Actions these days.

In this manner, the discussion over at https://github.com/pydata/numexpr/issues/367 has progressed and you might want to look into what @robbmcleod and @mattip (thank you!) have been doing / are suggesting there.

Specifically, [cibuildwheel](https://github.com/joerick/cibuildwheel) by @joerick, @YannickJadoul and contributors seems to be a nice way to build wheels these days and supports the full stack of Python versions and CI platforms.

With kind regards,
Andreas.

**Comment by avalentino** [comment]
Dear Andreas,
thank you for the pointers. I also used cibuildwheel to build wheels locally. I agree with you, it is a very nice tool.
I have just uploaded wheels for GNU/Linux.
Unfortunately I do not have too much time in this moment to setup a CI job to generate and upload wheels automatically.
If you are able to generate missing wheels please feel free to share them in some way or contact me in private  and I will upload them to PyPi.
In any case if you are interested to collaborate with the PyTables project I think we can give you commit access to the repo.

**Comment by amotl** [comment]
Dear Antonio,

well, I will be happy to provide a wheel for macOS using Python 3.9. Saying that, I am a newcomer here so I will probably have to learn from you first.

I just familiarized myself a little bit with the matter by cloning the repository, building it and running the tests on Python 3.8.6. It worked fine so far.

Now, I ran a build using `python setup.py bdist_wheel` within a pyenv'ed Python 3.9.0 virtualenv after switching to the `v3.6.1` tag. The outcome is [tables-3.6.1-cp39-cp39-macosx_10_13_x86_64.whl](https://filetransfer.io/data-package/htV8QGf6/download).

Do you believe using the internal blosc sources (see below) is fine?

With kind regards,
Andreas.

---

```
$ pip install cython numpy
Collecting cython
  Using cached Cython-0.29.21-py2.py3-none-any.whl (974 kB)
Collecting numpy
  Using cached numpy-1.19.4-cp39-cp39-macosx_10_9_x86_64.whl (15.4 MB)
Installing collected packages: cython, numpy
Successfully installed cython-0.29.21 numpy-1.19.4
```

---

```
$ python setup.py bdist_wheel
* Using Python 3.9.0 (default, Nov 17 2020, 23:27:37)
* USE_PKGCONFIG: True
* Found HDF5 headers at ``/usr/local/include``, library at ``/usr/local/lib``.
* Found LZO 2 headers at ``/usr/local/include``, library at ``/usr/local/lib``.
* Skipping detection of LZO 1 since LZO 2 has already been found.
* Found bzip2 headers at ``/usr/include``, library at ``/usr/lib``.
/var/folders/06/w9pzygdj7vx53n_0l9q_lhph0000gn/T/blosc_list_compressors2dldxok7.c:2:5: warning: implicit declaration of function 'blosc_list_compressors' is
      invalid in C99 [-Wimplicit-function-declaration]
    blosc_list_compressors();
    ^
1 warning generated.
ld: library not found for -lblosc
clang: error: linker command failed with exit code 1 (use -v to see invocation)
* Could not find blosc headers and library; using internal sources.
```

---

```
$ python -m tables.tests.test_all
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
PyTables version:    3.6.1
HDF5 version:        1.12.0
NumPy version:       1.19.4
Numexpr version:     2.7.1 (not using Intel's VML/MKL)
Zlib version:        1.2.11 (in Python interpreter)
LZO version:         2.10 (Mar 01 2017)
BZIP2 version:       1.0.6 (6-Sept-2010)
Blosc version:       1.16.3 (2019-03-08)
Blosc compressors:   blosclz (1.1.0), lz4 (1.8.3), lz4hc (1.8.3), snappy (1.1.1), zlib (1.2.8), zstd (1.3.8)
Blosc filters:       shuffle, bitshuffle
Cython version:      0.29.21
Python version:      3.9.0 (default, Nov 17 2020, 23:27:37)
[Clang 10.0.0 (clang-1000.10.44.4)]
Platform:            macOS-10.13.6-x86_64-i386-64bit
Byte-ordering:       little
Detected cores:      8
Default encoding:    utf-8
Default FS encoding: utf-8
Default locale:      (None, UTF-8)
```

**Comment by amotl** [comment]
Hi again,

> Unfortunately I do not have too much time in this moment to setup a CI job to generate and upload wheels automatically.

I am in the same situation, but maybe I can spare some time for that. At https://github.com/pydata/numexpr/issues/367#issuecomment-727765557, @mattip shared some recommendations.

While he said that NumPy uses Travis CI, https://github.com/numpy/numpy/pull/17738 shows that all Linux x86 jobs moved to GHA already and they are also planning to do this for the Windows/macOS builds.

Should PyTables go the same route and maybe already use [cibuildwheel](https://github.com/joerick/cibuildwheel) for that job?

With kind regards,
Andreas.

**Comment by mattip** [comment]
Just to clarify: my intention in the comment you linked to was to point out that NumPy still uses travis-ci for some non-x86 platforms.

**Comment by amotl** [comment]
All right, thanks for clarifying this, Matti! I believe that ARM64/AArch64 will gain more popularity in the future and I am looking forward to learn how to build packages for this by reusing what you are doing at NumPy [1].

[1] https://github.com/numpy/numpy/blob/master/.travis.yml

**Comment by YannickJadoul** [comment]
Not a lot of times to get into a library/build process that I don't know, but if there are questions about `cibuildwheel`, happy to answer them :-) (here or on Gitter, or wherever suits best)

Apart from the cibuildwheel docs, I'd suggest just starting out with [the examples](https://github.com/joerick/cibuildwheel/tree/master/examples) and seeing what goes wrong. If `setup.py bdist_wheel` or `pip wheel .` don't need too many weird dependencies or custom steps to prepare, things should be fairly smooth.

**Comment by amotl** [comment]
Thanks for your encouragement, Yannick!

As the whole build machinery of @NumPy looks rather complex, I believe it might not yet be possible to easily go along with `cibuildwheel`. However, here with `PyTables`, I just had to invoke `pip install cython numpy` before being able to build with `python setup.py bdist_wheel`.

So, I will be happy to apply the `cibuildwheel` GHA example here if @avalentino does not have any objections.

**Comment by amotl** [comment]
Dear @mattip,

accidentally, I just discovered how SQLAlchemy runs tests and builds wheels for ARM64/AArch64 on GHA through Docker/QEMU, see https://github.com/sqlalchemy/sqlalchemy/pull/5480 by @odidev.

With kind regards,
Andreas.

**Comment by amotl** [comment]
Dear @YannickJadoul and @mattip,

through https://github.com/tornadoweb/tornado/pull/2923, I also just discovered the [Run-On-Arch GitHub Action](https://github.com/marketplace/actions/run-on-architecture), which promises to execute commands on armv6, armv7, aarch64, s390x and ppc64le.

However, @bdarnell said at https://github.com/tornadoweb/tornado/pull/2923#issuecomment-699570811 that

> the sticking point is that `cibuildwheel` doesn't know how to build `arm64` wheels anywhere but travis-ci.

Maybe this is something you could look into, Yannick?

With kind regards,
Andreas.

**Comment by avalentino** [comment]
Dear @amotl personally I think that GHA + cibuildwheel is the way to go.
Don't know if other developers have comments about this.

The key point for PyTables builds is the setup of C/C++ libraries necessary to build.
Blosc is not a problem because we have the source in the repo but:

* hdf5
* zlib
* bzip2 (optional)
* lzo2 (optional)

must be installed on the system before building wheels.

Fro linux I just used 
```
CIBW_BEFORE_ALL_LINUX="yum install -y hdf5-devel zlib-devel bzip2-devel lzo-devel"
```

For osx it seems to me that [pytables-wheels](https://github.com/MacPython/pytables-wheels) uses brew,
Not sure what happens for windows builds.

My recommendation is to go by steps: setting up the base PR with linux only (please mark it as [WIP]), and then adding other OSs and architectures gradually (eventually also in separate PRs).

Thanks a lot for your effort.

**Comment by avalentino** [comment]
Very nice!
Thanks for the pointers @amotl.

**Comment by YannickJadoul** [comment]
@amotl

> However, here with `PyTables`, I just had to invoke `pip install cython numpy` before being able to build with `python setup.py bdist_wheel`.

That should be pretty straightforward to do with `CIBW_BEFORE_BUILD`, indeed.

The mentioned C libraries are potentially a bit more tricky, but should still work similarly. The issue might be on how to package these into a wheel for Windows (since Windows doesn't have an `auditwheel` or `delocate` variant (yet?); related to joerick/cibuildwheel#459 as well). Any idea how that happens currently, to provide Windows wheels?

> through [tornadoweb/tornado#2923](https://github.com/tornadoweb/tornado/pull/2923), I also just discovered the [Run-On-Arch GitHub Action](https://github.com/marketplace/actions/run-on-architecture) by @uraimo, which promises to execute commands on armv6, armv7, aarch64, s390x and ppc64le. At its core, it also uses [non-x86 emulation in Docker via QEMU](https://github.com/uraimo/run-on-arch-action/blob/v2.0.7/src/run-on-arch.sh#L34-L41).

That's pretty cool, thanks! Maybe this is worth opening a cibuildwheel issue for (or add this to an already existing issue, like joerick/cibuildwheel#364?). It should not be all too hard to enable `cibuildwheel` for this, I think. However, I would (as @avalentino rightfully suggests) take this bit by bit, and start out with the basic GHA example and adapt it gradually to build PyTables wheels. Already having manylinux, macOS, and Windows wheels for Intel architecture seems to be a big step forward?

**Comment by amotl** [comment]
Dear Antonino and Yannick,

> However, I would (as @avalentino rightfully suggests) take this bit by bit, and start out with the basic GHA example and adapt it gradually to build PyTables wheels. Already having manylinux, macOS, and Windows wheels for Intel architecture seems to be a big step forward?

Sure, sorry for hijacking this thread a bit. But there have been so much valuable information coming together with nice people caring about these topics (thanks!), so I believed it to be a good idea to share these findings with you within this very context.

> That's pretty cool, thanks! Maybe this is worth opening a `cibuildwheel` issue for (or add this to an already existing issue, like https://github.com/joerick/cibuildwheel/issues/364?).

Sure, will do.

> Personally, I think that GHA + cibuildwheel is the way to go.

Now, I will actually have to find some spare time to work on the low hanging fruits into the direction we are aiming at ;].

With kind regards,
Andreas.

**Comment by avalentino** [comment]
FYI I have setup a [basic GHA workflow](https://github.com/PyTables/PyTables/blob/master/.github/workflows/wheels.yml) for building wheels on linux. It would be nice to have also builds for windows and macosx.

**Comment by amotl** [comment]
Hi again,

based on @avalentino's b71ca545 (thanks!), I wanted to let you know that I am currently working on adding build targets for macOS and Windows on behalf of [1].

With kind regards,
Andreas.

[1] https://github.com/PyTables/PyTables/compare/master...daq-tools:gha-macos-windows-ci

**Comment by avalentino** [comment]
Hi @amotl, thanks a lot for the fantastic news!
If you prefer you can also open a PR and mark it as WIP until you complete your cleanup job.
Of course we will start reviewing only once you mark the PR as ready for merging.

**Comment by amotl** [comment]
Hi Antonio,

sure, I will open a PR soon, just fiddling with some more details. As I try to address the full procedure, may I humbly ask if you can add my handle to be able to upload packages to `test.pypi.org`?

> 403 The user 'amotl' isn't allowed to upload to project 'tables'. See https://test.pypi.org/help/#project-name for more information.

With kind regards,
Andreas.

**Comment by avalentino** [comment]
Hi @amotl not sure I can.
It seems that I'm not even registered to test.pypi.org (my standard pypi.org account doesn't work) so I doubt that I can control the pytables project on test.pypi.org.
Maybe @tomkooij or @FrancescAlted can.

**Comment by FrancescAlted** [comment]
Hi.  As far as I know, we don't have an account in test.pypi.org (I have tried my regular pypi.org credentials there, but it it is telling me that my username is not found).  Maybe @amotl can create a project with some name for testing purposes (pytables-test?) so as to test the uploading/downloading of wheels.

For what is worth, we recently added the build of wheels for python-blosc using the GitHub Actions infrastructure, and we are using GHA secrets in order to set the credentials for pypi.org (see PR https://github.com/Blosc/python-blosc/pull/237 for details).  My suggestion would be to do a similar thing for PyTables and use test.pypi.org initially (but not even this is needed, as you can download the wheels produced from GHA directly for local testing). When you would be happy, just get back to us and we will update the credentials for uploading to the real pypi.org.

**Comment by amotl** [comment]
Hi again,

I have been able to make some progress and issued a draft PR at #872.

As you can see from https://github.com/PyTables/PyTables/pull/872#issuecomment-770452079 - and as outlined at [1,2] - there is one thing missing yet. The build process croaks when trying to build wheels on Python 3.9/Windows with `fatal error LNK1181: cannot open input file 'hdf5.lib'`.

I can see the `setup.py` already does many sophisticated things in order to get things straight in this regard. However, on this specific case, it fails. Maybe one of the authors of these routines, eventually owning a Windows machine with a respective development environment can have a look? Maybe something like 29481179 (#540) by @tadeu or some adjustments to that is needed?

With kind regards,
Andreas.

cc @FrancescAlted, @avalentino, @eumiro, @andreabedini, @tomkooij, @robbmcleod, @tadeu

[1] https://github.com/PyTables/PyTables/blob/1acef23a79f54c404549c147cfc9e403f3f8e195/.github/workflows/wheels.yml#L134
[2] https://github.com/PyTables/PyTables/blob/1acef23a79f54c404549c147cfc9e403f3f8e195/.github/workflows/wheels.yml#L147

**Comment by avalentino** [comment]
Thanks a lot @amotl for the amazing job.
As far as I can understand it seems that many of the PyTables maintainers do not have an easy access to a windows machine. I don't know if there is an easy way to run a windows docker container on linux or osx.
Probably we can just go on and merge your PR #872 and leave the remaining wondows/python 3.9 part for a future PR.

P.S. How do you solved the problem with test.pypi.org?

**Comment by amotl** [comment]
Hi Antonio,

sure, let's go ahead and merge this PR and then look at what it's left later. @FrancescAlted: I could well use `pytables-test` for a testing upload to PyPI. However, I would rather take the chance in order to improve the situation with the original package.

So, I had a look: It looks like @tomkooij is the sole owner of the PyTables package on `test.pypi.org` [1]. It would be super nice if you, Tom, could add more people as owners to this package, maybe getting it en par with [2] by adding "falted", "antonio.valentino", "andreabedini", "jsancho" and "matthew.brett"?

Additionally, I will also be happy to be added on `test.pypi.org` [1], in order to test the PR #872 including its final step uploading the outcome to PyPI.

With kind regards,
Andreas.

[1] https://test.pypi.org/project/tables/
[2] https://pypi.org/project/tables/

**Comment by xmatthias** [comment]
Any news regarding the wheels for windows / macOS?

Not having these makes using tables (with python 3.9) on these platforms a real pain.

**Comment by tomkooij** [comment]
@amotl Ok, this is a VERY late reply. I just stumbled upon this.

I invited you (and antonio as he is the only owner with an account on pypi-test).

Sorry about not replying to this. Hope this still is usefull.

**Comment by avalentino** [comment]
Thanks @tomkooij.
You made a lot of work on wheel in the past.
Would you like to review this PR? Do you have time?

**Comment by pedrocamargo** [comment]
How is this issue coming along any help needed?

**Comment by joaoe** [comment]
Hi.
Just a friendly reminder that releases of python since 3.8 are not supported by pytables on mac nor windows, and you are forcing projects into technical debt by having to keep python 3.7.

**Comment by RalfG** [comment]
To people looking for an update on this issue, @xmatthias has done some great work already to get 3.9 and 3.10 wheels: 
- #917
- #918
- #923
- #926
- #927

Currently, open issues are:
- #929
- #930

**Comment by avalentino** [comment]
@xmatthias since #923 is now merged I think that this issue can be closed.
Please feel free to reopen if I missed something.

**Comment by joaoe** [comment]
@avalentino Hi. The Windows wheels are not published at https://pypi.org/project/tables/3.6.1/#files 
What is missing ? Thank you.

**Comment by avalentino** [comment]
The machinery to generate wheels is now fixed.
We are going to make a new release in few days.
